### PR TITLE
Add new HoloLens sample project - GazeRuler

### PR DESCRIPTION
**Pull request by ChangweiZhang**
GazeRuler is a project which can be used to measure spatial distance and arean. It's based on Gaze and Spatial Mapping and users can enjoy it by using voice command and gesture. 
I think that it's useful for some developers who want learn Spatial Mapping.

### CODE REVIEW

File: Assets/HoloToolkit-Examples/GazeRuler/README.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
+#What is GazeRuler
+
+GazeRuler is a project which can be used to measure spatial distance and arean. It's based on Gaze and Spatial Mapping and users can enjoy it by using voice command and gesture.
```
Comment (NeerajW) [inline Assets/HoloToolkit-Examples/GazeRuler/README.md]: What is arean exactly? Sorry I was unsure if this was a typo or not.

File: Assets/HoloToolkit-Examples/GazeRuler/README.md [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+#How to use?
+
+There are two measure modes in this app, including Line Mode and Geometry Mode. Line Mode is used to measure the distance between one spatial point and another point, when Geometry Mode is used to measure the area of spatial geometries.
+By default, Line Mode is on.
+![Line Mode](https://github.com/ChangweiZhang/HoloToolkit-Unity/blob/master/Assets/HoloToolkit-Examples/GazeRuler/Demo/Line.jpg)
+
+After you see the tip text of Line Mode, you can ait-tap to place one point and air-tap again to place another point. A line and tip of result will show automatically.
```
Comment (NeerajW) [inline Assets/HoloToolkit-Examples/GazeRuler/README.md]: NIT: Typo: air-tap

File: Assets/HoloToolkit-Examples/GazeRuler/README.md [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+![Geometry Mode](https://github.com/ChangweiZhang/HoloToolkit-Unity/blob/master/Assets/HoloToolkit-Examples/GazeRuler/Demo/Geometry.jpg)
+If you want try Geometry Mode, what's you only need to do is say "change mode". Once your voice command is recognized, this app will change into Geometry Mode.
+In Geometry Mode, you also can air-tap to place one point and air-tap again to place more points. When you want finish the geometry, please hold your handle or say "close". The geometry will be finished automatically and return the result of the area of current geometry.
+Of cause, you can go back to Line Mode by saying "change mode" again.
```
Comment (NeerajW) [inline Assets/HoloToolkit-Examples/GazeRuler/README.md]: NIT: Remove Of cause

File: Assets/HoloToolkit-Examples/GazeRuler/README.md [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+Last but not least, every line and geometry can be deleted by saying "remove" or "remove all".
```
Comment (NeerajW) [inline Assets/HoloToolkit-Examples/GazeRuler/README.md]: Do you need to gaze at a line to remove it?

File: Assets/HoloToolkit-Examples/GazeRuler/README.md [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+If you want find more details, please contact me: [http://changwei.tech](http://changwei.tech)
+
+for demo video, please click here: [http://player.youku.com/embed/XMTY2MTU1ODA5Ng==](http://player.youku.com/embed/XMTY2MTU1ODA5Ng==)
```
Comment (NeerajW) [inline Assets/HoloToolkit-Examples/GazeRuler/README.md]: There seems to be some localized character at the end of the Tooltip. Can you please fix that for en-us so it's more readable?

File: Assets/HoloToolkit-Examples/GazeRuler/Scripts/CWGestureManager.cs [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    }
+
+    private void OnTap()
+    {
+        //if user tap tip text, relative line and geometry will hide
+        if (focusedObject != null && focusedObject.tag.Equals("Tip"))
+        {
+            focusedObject.SendMessage("OnSelect");
+        }
+        else
```
Comment (NeerajW) [inline Assets/HoloToolkit-Examples/GazeRuler/Scripts/CWGestureManager.cs]: Please follow consistent coding guides. 
if (condition)
{
    // some code
}
else
{
    // some code
}

File: Assets/HoloToolkit-Examples/GazeRuler/Scripts/DeleteLine.cs [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+using System.Collections;
+
+public class DeleteLine : MonoBehaviour {
+    /// <summary>
+    /// when tip text is tapped, destroy this tip and relative objects.
+    /// </summary>
+	public  void OnSelect()
+    {
+        var parent = gameObject.transform.parent.gameObject;
+        if (parent != null)
```
Comment (NeerajW) [inline Assets/HoloToolkit-Examples/GazeRuler/Scripts/DeleteLine.cs]: Please follow curly braces pattern.

File: Assets/HoloToolkit-Examples/GazeRuler/Scripts/IClosable.cs [hunk 8]
```diff
@@ -0,0 +1,9 @@
+﻿using UnityEngine;
+
+/// <summary>
+/// any class inherit this interface should be closeable
+/// </summary>
+public interface IClosable  {
```
Comment (NeerajW) [inline Assets/HoloToolkit-Examples/GazeRuler/Scripts/IClosable.cs]: Did you mean to inherit from IClosable? Perhaps a better name may be? IPolygonClose or something?

File: Assets/HoloToolkit-Examples/GazeRuler/Scripts/IGeometry.cs [hunk 9]
```diff
@@ -0,0 +1,18 @@
+﻿using UnityEngine;
+using System.Collections;
+
+/// <summary>
+/// interface for geometry class
+/// </summary>
+public interface IGeometry
+{
+
```
Comment (NeerajW) [inline Assets/HoloToolkit-Examples/GazeRuler/Scripts/IGeometry.cs]: NIT: extra lines. Please remove.

File: Assets/HoloToolkit-Examples/GazeRuler/Scripts/LineManager.cs [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+
+            var directionFromCamera = centerPos - Camera.main.transform.position;
+
+            var distanceA = Vector3.Distance(lastPoint.Position, Camera.main.transform.position);
+            var distanceB = Vector3.Distance(hitPoint, Camera.main.transform.position);
+
+            Debug.Log("A: " + distanceA + ",B: " + distanceB);
+            Vector3 direction;
+            if (distanceB > distanceA || (distanceA > distanceB && distanceA - distanceB < 0.1))
+                direction = hitPoint - lastPoint.Position;
```
Comment (NeerajW) [inline Assets/HoloToolkit-Examples/GazeRuler/Scripts/LineManager.cs]: Please follow curly braces pattern for if else.

File: Assets/HoloToolkit-Examples/GazeRuler/Scripts/LineManager.cs [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+            else
+                direction = lastPoint.Position - hitPoint;
+
+            var distance = Vector3.Distance(lastPoint.Position, hitPoint);
+            var line = (GameObject)Instantiate(LinePrefab, centerPos, Quaternion.LookRotation(direction));
+            line.transform.localScale = new Vector3(distance, 0.005f, 0.005f);
```
Comment (NeerajW) [inline Assets/HoloToolkit-Examples/GazeRuler/Scripts/LineManager.cs]: Define the 0.005f into private const ?

File: Assets/HoloToolkit-Examples/GazeRuler/Scripts/LineManager.cs [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+
+            var normalV = Vector3.Cross(direction, directionFromCamera);
+            var normalF = Vector3.Cross(direction, normalV) * -1;
+
+
+            var tip = (GameObject)Instantiate(TextPrefab, centerPos, Quaternion.LookRotation(normalF));
+
+            //unit is meter
+            tip.transform.Translate(Vector3.up * 0.05f);
+            tip.GetComponent<TextMesh>().text = distance + "m";
```
Comment (NeerajW) [inline Assets/HoloToolkit-Examples/GazeRuler/Scripts/LineManager.cs]: I suspect this is what showed up as the character in your video :) You locale might be set different :)

File: Assets/HoloToolkit-Examples/GazeRuler/Scripts/MeasureManager.cs [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+        {
+            Debug.Log(ex.Message);
+        }
+
+        ModeTipObject.SetActive(true);
+        
+
+    }
+}
+
```
Comment (NeerajW) [inline Assets/HoloToolkit-Examples/GazeRuler/Scripts/MeasureManager.cs]: Please do go through all your files and remove extra lines.

File: Assets/HoloToolkit-Examples/GazeRuler/Scripts/ModeTip.cs [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+    
+    void Start()
+    {
+        text = GetComponent<TextMesh>();
+
+        switch (MeasureManager.Instance.mode)
+        {
+
+            case GeometryMode.Line:
+                text.text = "Line测距模式";
```
Comment (NeerajW) [inline Assets/HoloToolkit-Examples/GazeRuler/Scripts/ModeTip.cs]: Please have these localized to en-us.

File: Assets/HoloToolkit-Examples/GazeRuler/Scripts/PloygonManager.cs [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+                vect += point;
+            }
+            var centerPoint = vect / (index + 1);
+
+            var direction1 = CurrentPloygon.Points[1] - CurrentPloygon.Points[0];
+            var directionF = Vector3.Cross(direction, direction1);
+            var tip = (GameObject)Instantiate(TextPrefab, centerPoint, Quaternion.LookRotation(directionF));//anchor.x + anchor.y + anchor.z < 0 ? -1 * anchor : anchor));
+
+            // unit is ㎡
+            tip.GetComponent<TextMesh>().text = area + "㎡";
```
Comment (NeerajW) [inline Assets/HoloToolkit-Examples/GazeRuler/Scripts/PloygonManager.cs]: Or perhaps this character does not show up correctly?

**Comment by NeerajW** [comment]
@ChangweiZhang absolutely delighted to see your contributions. Exactly how we want to use the HoloToolkit-Examples folder :)

File: Assets/HoloToolkit-Examples/GazeRuler/README.md (hunk 1 above)
Comment (ChangweiZhang) [inline Assets/HoloToolkit-Examples/GazeRuler/README.md]: sorry, 'arean' should be 'area', and it's a error

File: Assets/HoloToolkit-Examples/GazeRuler/README.md (hunk 2 above)
Comment (ChangweiZhang) [inline Assets/HoloToolkit-Examples/GazeRuler/README.md]: thanks

File: Assets/HoloToolkit-Examples/GazeRuler/README.md (hunk 3 above)
Comment (ChangweiZhang) [inline Assets/HoloToolkit-Examples/GazeRuler/README.md]: thanks

File: Assets/HoloToolkit-Examples/GazeRuler/README.md (hunk 4 above)
Comment (ChangweiZhang) [inline Assets/HoloToolkit-Examples/GazeRuler/README.md]: User can gaze at the tip text of line and air-tap to remove the line

File: Assets/HoloToolkit-Examples/GazeRuler/README.md (hunk 5 above)
Comment (ChangweiZhang) [inline Assets/HoloToolkit-Examples/GazeRuler/README.md]: ok

File: Assets/HoloToolkit-Examples/GazeRuler/Scripts/CWGestureManager.cs (hunk 6 above)
Comment (ChangweiZhang) [inline Assets/HoloToolkit-Examples/GazeRuler/Scripts/CWGestureManager.cs]: thanks

File: Assets/HoloToolkit-Examples/GazeRuler/Scripts/DeleteLine.cs (hunk 7 above)
Comment (ChangweiZhang) [inline Assets/HoloToolkit-Examples/GazeRuler/Scripts/DeleteLine.cs]: thanks

File: Assets/HoloToolkit-Examples/GazeRuler/Scripts/IClosable.cs (hunk 8 above)
Comment (ChangweiZhang) [inline Assets/HoloToolkit-Examples/GazeRuler/Scripts/IClosable.cs]: good suggestion, thanks

File: Assets/HoloToolkit-Examples/GazeRuler/Scripts/LineManager.cs (hunk 10 above)
Comment (ChangweiZhang) [inline Assets/HoloToolkit-Examples/GazeRuler/Scripts/LineManager.cs]: thanks

File: Assets/HoloToolkit-Examples/GazeRuler/Scripts/LineManager.cs (hunk 11 above)
Comment (ChangweiZhang) [inline Assets/HoloToolkit-Examples/GazeRuler/Scripts/LineManager.cs]: ok

File: Assets/HoloToolkit-Examples/GazeRuler/Scripts/LineManager.cs (hunk 12 above)
Comment (ChangweiZhang) [inline Assets/HoloToolkit-Examples/GazeRuler/Scripts/LineManager.cs]: yep, and I will fix it for en-us

File: Assets/HoloToolkit-Examples/GazeRuler/Scripts/MeasureManager.cs (hunk 13 above)
Comment (ChangweiZhang) [inline Assets/HoloToolkit-Examples/GazeRuler/Scripts/MeasureManager.cs]: thanks

**Comment by ChangweiZhang** [comment]
Hi NeerajW, thank you for you patient and I will fix all errors as soon as possible

**Comment by ChangweiZhang** [comment]
@NeerajW I have fix my source code and README file

File: Assets/HoloToolkit-Examples/GazeRuler/Scripts/Test/LineTest.cs [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+/// </summary>
+public class LineTest : MonoBehaviour
+{
+
+
+    public GameObject start;
+    public GameObject end;
+    public GameObject line;
+    public GameObject text;
+    private bool created = false;
```
Comment (jwittner) [inline Assets/HoloToolkit-Examples/GazeRuler/Scripts/Test/LineTest.cs]: Assets/HoloToolkit-Examples/GazeRuler/Scripts/Test/LineTest.cs(15,18): warning CS0414: The private field `LineTest.created' is assigned but its value is never used

**Comment by jwittner** [comment]
Other than the warning I just commented on, things look pretty great! Ran it in the emulator and found it pretty intuitive!

**Comment by jwittner** [comment]
@NeerajW Did you just merge this in with that warning? Ok if you did, I'll just make an issue about it.


Assets/HoloToolkit-Examples/GazeRuler/Scripts/Test/LineTest.cs(15,18): warning CS0414: The private field `LineTest.created' is assigned but its value is never used

**Comment by NeerajW** [comment]
@jwittner you are fast my friend :) Done #153.

**Pull request merged**